source/testing/index.rst

```diff
@@ -0,0 +1,499 @@
+database contents, and that models validate, save, and are properly
+updated by other helpers. All of these require tests that actually
+work on a real database, because some behaviors can only be replicated
+accurately with a full round trip of the SQL.
+
+The Django test runner provides facilities that create a new database
+just for the testsand run migrations to create all the tables your
+applications need, and your test suites can create and remove test
+data in their ``setUp`` and ``tearDown`` methods. This can be
+invaluable, but also comes at a cost in test run time.
+
+Inheriting from ``django.test.TestCase`` will run your tests in a
+database transaction, and require the test database building, in order
+to run the tests. Tests directly using ``unittest.TestCase``, however,
+will run like any other Python unittests and avoid these extra
+database tests. Distinguishing which test cases you need the database
+support for can help you speed up your test run time, and also better
+frame which areas of your application are responsible for database
+behaviors and which are not.
+
+Be aware, though, that if database access creeps into a test that
+isn't using ``django.test.TestCase``, it will make permanent
+changes to the test database and often break unrelated tests that
+assume a pristine database.  This can be very hard to debug.
+It's a good idea to add some protection
+against this. For example, see `slide 29 of Carl Meyer's talk
+on django testing <http://carljm.github.io/django-testing-slides/#29>`_
+for a way to immediately cause tests to fail if they access the
+database when they weren't expected to.
+
+A note on data fixtures: Django provides a feature for adding test
+data from JSON or other formats, but in practice we’ve found this to
+be an impractical tool and advise against it. Problems include
+difficulties updating your fixtures when data schemas and requirements
+change; and the disconnect between reading your test code and your
+test data. Instead, tests that create data programmatically in the
+setUp() or test methods will give you tests that are easier to read,
+easier to setup, and in most cases continue to function during schema
+changes.
+
+References:
+
+* `Django: Writing and running tests <https://docs.djangoproject.com/en/stable/topics/testing/overview/>`_
+* Carl J Meyer: Testing and Django `slides <http://carljm.github.com/django-testing-slides/>`_
+  and `video <http://www.youtube.com/watch?v=ickNQcNXiS4>`_. Unfortunately,
+  the first part (slides 2-12) is
+  out-of-date with behavior of current versions of Django (1.9 as I write this), so
+  skip ahead to slide 13 and continue from there.
+
+How to test Django template tags
+++++++++++++++++++++++++++++++++
+
+Testing template tags is an often overlooked and difficult aspect of
+testing in Django apps, but if you break up the logic right, you can
+actually test your custom tags easily.
+
+As an example, testing separately the parsing of a tag::
```

Comment:
This might be just my inexperience with writing template tags, but I think showing the code being tested (or maybe just the an example of the template tag being used, like Calvin has on his blog) might help me understand these examples a bit more.